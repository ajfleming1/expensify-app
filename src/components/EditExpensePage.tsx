import React from "react";
import { RouteComponentProps } from "react-router-dom";
type IdParams = { id: string };

const EditExpensePage = (props: RouteComponentProps<IdParams>) => 
{
    console.log(props);
    return (<div>Editing the expense with id {props.match.params.id}.</div>);
}

export default EditExpensePage;