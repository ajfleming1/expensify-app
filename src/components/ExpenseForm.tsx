import React, { ChangeEvent, FormEvent } from "react";
import moment, { Moment } from "moment";
import { SingleDatePicker } from "react-dates";
import { v4 as uuid } from 'uuid';
import 'react-dates/initialize';
import { UpdateExpenseType, ExepenseItemType } from "../@types/expenseTypes";

class ExpenseForm extends React.Component<{ expense?: ExepenseItemType, onSubmit: (expense: UpdateExpenseType) => void }> {
  state = {} as {
    description: string,
    note: string,
    amount: string,
    createdAt: Moment,
    calendarFocused: boolean,
    error: string
  };

  constructor(props: { expense: ExepenseItemType; onSubmit: (expense: UpdateExpenseType) => void; }) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ""
    };
  }

  onDescriptionChange = (e: ChangeEvent<HTMLInputElement>) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  }

  onNoteChange = (e: ChangeEvent<HTMLTextAreaElement>) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = (e: ChangeEvent<HTMLInputElement>) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = (createdAt: moment.Moment) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }: any) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: "Please provide an amount and a description." }));
    }
    else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit(
        {
          description: this.state.description,
          amount: parseFloat(this.state.amount) * 100,
          createdAt: this.state.createdAt.valueOf(),
          note: this.state.note
        })
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            id={uuid()}
            numberOfMonths={1}
            isOutsideRange={() => (false)}
          />
          <textarea
            placeholder="Add a note for your expense (optional)."
            value={this.state.note}
            onChange={this.onNoteChange} >
          </textarea>
          <button>{this.props.expense ? "Update Expense" : "Add Expense"}</button>
        </form>
      </div>
    )
  }
};

export default ExpenseForm;