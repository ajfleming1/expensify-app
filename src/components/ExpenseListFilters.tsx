import React from "react";
import { Props, connector } from "../@types/rootState";
import { connect } from "react-redux";
import { setTextFilter } from "../actions/filters";
const ExpenseListFilters = (props: Props) => (
  <div>
    <input type="text" value={props.filters.text} onChange=
      {
        (e) => props.dispatch(setTextFilter(e.target.value))
      }
    />
  </div>
);

export default connector(ExpenseListFilters);