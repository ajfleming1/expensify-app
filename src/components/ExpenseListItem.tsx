import React from "react";
import { Expense } from "../@types/expenseTypes";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import moment from "moment";
import numerial from "numeral";

export const ExpenseListItem = (expense: Expense) => (
  <Link className="list-item" to={`/edit/${expense.id}`}>
    <div>
      <h3 className="list-item__title">
        {expense.description}
      </h3>
      <span className="list-item__subtitle">
        {moment(expense.createdAt).format("MMMM Do, YYYY")}
      </span>
    </div>
    <h3 className="list-item__data">
      {numerial(expense.amount / 100).format("$0,0.00")}
    </h3>
  </Link>
);

export default connect()(ExpenseListItem);