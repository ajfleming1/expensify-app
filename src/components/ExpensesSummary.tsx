import React from "react";
import selectExpensesTotal from "../selectors/expenses-total";
import { RootState } from "../@types/rootState";
import selectExpenses from "../selectors/expenses";
import numerial from "numeral";
import { connect } from "react-redux";

export const ExpensesSummary = ({ expenseCount, expensesTotal }: { expenseCount: number, expensesTotal: number }) => {
  const expenseWord = expenseCount === 1 ? "expense" : "expenses";
  return (
    <div>
      <h1>
        Viewing {expenseCount} {expenseWord} totalling {numerial(expensesTotal / 100).format("$0,0.00")}.
      </h1>
    </div>
  );
};

const mapStateToProps = (state: RootState) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses)
  }
};

const connector = connect(mapStateToProps);
export default connector(ExpensesSummary);