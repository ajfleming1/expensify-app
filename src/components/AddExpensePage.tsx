import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { UpdateExpenseType } from "../@types/expenseTypes";
import { Props } from "../@types/rootState";
import { addExpense } from "../actions/expenses";
import { Dispatch } from "redux";

export class AddExpensePage extends React.Component<Props & DispatchProps> {
  onSubmit = (expense: UpdateExpenseType) => {
    this.props.onSubmit(expense);
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <h1>Add Expense </h1>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}

interface DispatchProps {
  onSubmit: (expense: UpdateExpenseType) => void
}

const mapDispatchToProps = (dispatch: Dispatch) => ({
  onSubmit: (expense: UpdateExpenseType) => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);