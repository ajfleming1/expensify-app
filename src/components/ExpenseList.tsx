import React from "react";
import { Props, connector } from "../@types/rootState";
import ExpenseListItem from "./ExpenseListItem";
import { Expense } from "../@types/expenseTypes";

export const ExpenseList = (props: Props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-body">
      {
        props.expenses.length === 0 ? (
          <div className="list-item list-item--message"><span>No Expenses</span></div>
        ) : (
            props.expenses.map((expense: Expense) => {
              return <ExpenseListItem key={expense.id} {...expense} />
            })
          )
      }
    </div>
  </div>
);

export default connector(ExpenseList);
