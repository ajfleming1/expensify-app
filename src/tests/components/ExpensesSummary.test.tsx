import React from "react";
import { shallow } from "enzyme";
import { ExpensesSummary } from "../../components/ExpensesSummary";

test("should have correct output for two items", () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={195} />);
  expect(wrapper).toMatchSnapshot();
});

test("should have correct output for two items", () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={10} expensesTotal={1124566} />);
  expect(wrapper).toMatchSnapshot();
});