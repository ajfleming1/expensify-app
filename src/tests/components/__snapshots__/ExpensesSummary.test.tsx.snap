// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should have correct output for two items 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $1.95
    .
  </h1>
</div>
`;

exports[`should have correct output for two items 2`] = `
<div>
  <h1>
    Viewing 
    10
     
    expenses
     totalling 
    $11,245.66
    .
  </h1>
</div>
`;
